### PR TITLE
refactor: Make `SpaceRoom` compute `is_dm` on demand

### PR DESCRIPTION
**Pull request by jmartinesp**
This means making both `SpaceRoom::new_from_summary` and `SpaceRoom::new_from_known_room` async, which in turn forces us to change how we instantiate those in quite a few places.

Why do we need this? It seems like the pre-computed `active_service_member_count` is not usually available for these rooms: if you are in spaces with lots of rooms you haven't visited before, you wouldn't have triggered their `sync_members` call that would compute this value. In those cases, you'd get the wrong computed room display names and `is_dm` values.

Also, I tweaked a bit how the display name computation and the `heroes` values are returned:

- The computed display name will try computing the joined service member count first now that we can do it, so that rooms with service members display the proper display name (i.e. in a room with you, Alice and a Bot service member it'll display simply 'Alice' instead of 'Alice, and 2 others', those 2 others being you and the bot). This happens because in `compute_display_name_from_heroes` the `if num_heroes < num_joined_invited_except_self && num_joined_invited > 1` line would always be triggered for rooms with service members.
- When using the heroes to compute the display name and when returning the `heroes` field in `SpaceRoom`, use `Room::heroes` instead of `RoomInfo::heroes`. The latter just stores what the HS returns, while the former filters out service members that should not be considered heroes.

---

Changes after review: the `SpaceRoomList::rooms` Mutex is now an `AsyncMutex`. This changes some other method signatures so they're async now too.

- [x] I've documented the public API Changes in the appropriate `CHANGELOG.md` files.
- [ ] This PR was made with the help of AI.

Signed-off-by:

### CODE REVIEW

File: crates/matrix-sdk-ui/src/spaces/mod.rs [hunk 1]
```diff
@@ -657,6 +656,23 @@ impl SpaceService {
     }
 }
 
+async fn children_rooms(client: &Client, children: &[OwnedRoomId], graph: &SpaceGraph) -> Vec<SpaceRoom> {
```
Comment (stefanceriu) [inline crates/matrix-sdk-ui/src/spaces/mod.rs]: Should we call this `children_space_rooms` to make things just a bit clearer? Or maybe use `children_ids`, or both?

File: crates/matrix-sdk-ui/src/spaces/room_list.rs [hunk 2]
```diff
@@ -144,24 +144,32 @@ impl SpaceRoomList {
                                     continue;
                                 }
 
-                                let mut mutable_rooms = rooms.lock();
-
-                                updates.iter_all_room_ids().for_each(|updated_room_id| {
+                                let mut to_set = Vec::new();
```
Comment (stefanceriu) [inline crates/matrix-sdk-ui/src/spaces/room_list.rs]: I initially thought you're going to convert this to a Set eventually. Can we rename it to `updated` or something?

File: crates/matrix-sdk-ui/src/spaces/room_list.rs [hunk 3]
```diff
@@ -144,24 +144,32 @@ impl SpaceRoomList {
...
+                                for updated_room_id in updates.iter_all_room_ids() {
+                                    let mutable_rooms = rooms.lock();
```
Comment (stefanceriu) [inline crates/matrix-sdk-ui/src/spaces/room_list.rs]: Do we really need to aquire and drop this with every loop?

**Comment by jmartinesp** [comment]
I think I handled all the feedback in https://github.com/matrix-org/matrix-rust-sdk/pull/6561/commits/48740d1f821afd5d7451997f6a887e5f5e9a6f34, let me know if something's missing.

File: crates/matrix-sdk-ui/src/spaces/room_list.rs (hunk 3 above)
Comment (jmartinesp) [reply crates/matrix-sdk-ui/src/spaces/room_list.rs]: Looks like we don't, as long as everything is in its own block so the compiler understands the lock is no longer in use (it seems like manually `drop`ing it wasn't enough).

**Review by stefanceriu — APPROVED** [review]
The test are not happy, but I am, 🐑🚀 it!

**Comment by jmartinesp** [comment]
Many thanks to @frebib for his help while testing this, including a corner case I completely missed 🙏 .
